File: 0141-linked-list-cycle/0141-linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (!head || !head->next) {
            // If the list is empty or has only one node, there can't be a cycle
            return false;
        }
        
        ListNode *tortoise = head;
        ListNode *hare = head;
        
        while (hare && hare->next) {
            tortoise = tortoise->next;
            hare = hare->next->next;
            
            // If there is a cycle, the tortoise and hare will meet at some point
            if (tortoise == hare) {
                return true;
            }
        }
        
        // If the hare reaches the end of the list, there is no cycle
        return false;
    }
};

File: 0141-linked-list-cycle/0141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        self.slow=head #create a variable name slow point at head
        self.fast=head #same with name fast
        while self.fast is not None and self.fast.next is not None:
            #now check the condition again that if nodes are never get null then it must be in cycle
            self.fast = self.fast.next.next
            #where fast variable increment by two nodes and slow with one 
            self.slow=self.slow.next
            # there is a method floyed cycle finding alogrithm ,where if both are in same place then we can say that this are in loop
            if self.fast==self.slow:
                return True
            # and return true 
        return False  
    # else retrun False
        